driver-core/src/main/java/com/datastax/driver/core/PagingState.java

```diff
@@ -0,0 +1,176 @@
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package com.datastax.driver.core;
+
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+import com.datastax.driver.core.exceptions.PagingStateException;
+import com.datastax.driver.core.utils.Bytes;
+
+/**
+ * The paging state of a query.
+ *
+ * This object represents the next page to be fetched if the query is
+ * multi page. It can be saved and reused later on the same statement.
+ *
+ * The PagingState can be serialized and deserialized either in a String
+ * or in a ByteBuffer.
+ */
+public class PagingState {
+
+    private ByteBuffer pagingState;
+
+    PagingState(ByteBuffer pagingState, Statement statement) {
+        this.pagingState = generatePagingState(pagingState, hash(statement, pagingState));
+    }
+
+    private PagingState(ByteBuffer pagingState) {
+        this.pagingState = pagingState;
+    }
+
+    private ByteBuffer hash(Statement statement, ByteBuffer pagingState) {
+        byte[] digest;
+        ByteBuffer[] values;
+        MessageDigest md;
+
+        assert !(statement instanceof BatchStatement);
+
+        try {
+            md = MessageDigest.getInstance("MD5");
+            if (statement instanceof BoundStatement) {
+                BoundStatement bs = ((BoundStatement)statement);
+                md.update(bs.preparedStatement().getQueryString().getBytes());
+                values = bs.values;
+            } else {
+                //it is a RegularStatement since Batch statements are not allowed
+                RegularStatement rs = (RegularStatement)statement;
+                md.update(rs.getQueryString().getBytes());
+                values = rs.getValues();
+            }
+            if (values != null) {
+                for (ByteBuffer value : values) {
+                    md.update(value.duplicate());
+                }
+            }
+            md.update(pagingState.duplicate());
+            digest = md.digest();
+
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("MD5 doesn't seem to be available on this JVM", e);
+        }
+        return ByteBuffer.wrap(digest);
+    }
+
+    boolean matches(Statement statement) {
+        this.pagingState.rewind();
+        // Check the sizes
+        int pagingSize = this.pagingState.getShort();
+        int hashSize = this.pagingState.getShort();
+        if (pagingSize + hashSize != this.pagingState.remaining()) {
+            return false;
+        }
+        ByteBuffer hashToCheck = extractHashDigest(this.pagingState);
+        ByteBuffer extractedPagingState = extractPagingState(this.pagingState);
+        ByteBuffer toTest = this.hash(statement, extractedPagingState);
+        return hashToCheck.equals(toTest);
+    }
+
+    private ByteBuffer extractPagingState(ByteBuffer complete) {
```

Comment:
I've been thinking about that (I am currently making the changes to match your remark on separating the paging state and the hash, so it changes the structure) , and that would make sense but, now that i have made the changes I was thinking about a method like "extract" that would extract the paging state byte array from a byte array generated with toByte() (hence, get the size, then the paging state part). And another method getRawState() to just get the paging state buffer of the object (wrapped into a ByteBuffer because it is used as a ByteBuffer further in the driver). Is that ok to do it like that ?